tools/enum_macro/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+                .iter()
+                .find(|attr| attr.path().is_ident("permission"))
+            {
+                if let syn::Meta::List(l) = &attr.meta {
+                    if let Err(err) = l.parse_nested_meta(|meta| {
+                        if meta.path.is_ident("formatting") {
+                            let exp: syn::Expr = meta.value()?.parse()?;
+                            if let syn::Expr::Lit(v) = exp {
+                                if let syn::Lit::Str(s) = v.lit {
+                                    format = Some(s.value());
+                                }
+                            }
+                            return Ok(());
+                        }
+                        Err(meta.error("Unrecognized attribute"))
+                    }) {
+                        panic!("{:?}", err);
+                    }
+                }
+            }
+            let format = format.unwrap_or("{}".into());
+
+            let variant_name = &variant.ident;
+
+            variants.push(quote! {
+                #name::#variant_name => {
+                    let mut snake = String::new();
+                    for (i, ch) in stringify!(#variant_name).char_indices() {
+                        if i > 0 && ch.is_uppercase() {
+                            snake.push('_');
+                        }
+                        snake.push(ch.to_ascii_lowercase());
+                    }
+                    format!(#format, snake)
+                },
+            });
+        }
+
+        let expanded = quote! {
+            impl std::fmt::Display for #name {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    let result = match self {
+                        #(#variants)*
+                    };
+                    f.write_str(&format!("{}", result))
+                }
+            }
+        };
+
+        TokenStream::from(expanded)
+    } else {
+        panic!("Only Enums are supported");
+    }
+}
+
+/// Implements #values() for enum values that returns all defined values in the enum
+/// E.g.
+/// ```
+/// use enum_macro::EnumValues;
+/// #[derive(EnumValues, PartialEq, Debug)]
+/// pub enum SomeEnum {
+///     SomeValue1,
+///     SomeValue2
+/// }
+///
+/// assert_eq!([SomeEnum::SomeValue1, SomeEnum::SomeValue2], SomeEnum::values());
+/// ```
+#[proc_macro_derive(EnumValues)]
+pub fn enum_values(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+    if let Data::Enum(enum_data) = input.data {
+        let variants: Vec<_> = enum_data.variants.into_iter().map(|v| v.ident).collect();
+        let expanded = quote! {
+            impl #name {
+                pub fn values() -> &'static[#name] {
+                    &[ #(#name::#variants),* ]
+                }
+            }
+        };
+        TokenStream::from(expanded)
+    } else {
+        panic!("Only Enums are supported");
+    }
+}
```

Comment:
Wouldnt it be better to just use serde to rename the enum than to use a macro?

Reply:
serde can only to something like upper/lowercase etc. they dont have anything like formatting sadly

Reply:
you can use `#[serde(rename="some.permission.node")]` for each enum variant. Ie

```rust
#[derive(Serialize, Deserialize)]
enum PermissionNode {
    #[serde(rename="add.identity")]
    AddIdentity,
    #[serde(rename="remove.identity")]
    RemoveIdentity
}
```

Reply:
yes but with that its a bit decoupled from the actual enum name which i didnt like

Reply:
Thats only an example though. It could be named anything, however serde would handle that though when it comes to deserializing it back to the enum.

Reply:
i meant that serde does not give access to the enum name inside renaming. so if you e.g. rename the enum itself `AddIdentity` you also need to update the serialized name. its a bit nitpicky but initially i wanted the enums purely for sanity checks within warp and allow arbitrary strings as permissions

Reply:
In that case you would want to serialize it manually if you dont want to update the rename attr macro, although this shouldnt exactly be a problem to rename it either (though to take caution against breaking changes too).

Reply:
i can change it back to serde if you think its better. i personally think manually defining the string everytime is a bit cumbersome